pq-crypto/s2n_kyber_512_evp.c

```diff
@@ -20,13 +20,14 @@
 
 #include "error/s2n_errno.h"
 #include "tls/s2n_kem.h"
+#include "utils/s2n_safety.h"
 #include "utils/s2n_safety_macros.h"
 
 #if defined(S2N_LIBCRYPTO_SUPPORTS_KYBER512)
 int s2n_kyber_512_evp_generate_keypair(uint8_t *public_key, uint8_t *private_key) {
-    EVP_PKEY_CTX *kyber_pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_KYBER512, NULL);
+    DEFER_CLEANUP(EVP_PKEY_CTX *kyber_pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_KEM, NULL), EVP_PKEY_CTX_free);
```

Comment:
This shouldn't compile. The "free" method you give to DEFER_CLEANUP has to take a pointer to the type you call it on, so here it would need to take EVP_PKEY_CTX**, not EVP_PKEY_CTX*. 

You need to be checking for LIBCRYPTO_SUPPORTS_EVP_KEM_KYBER_512 now instead of S2N_LIBCRYPTO_SUPPORTS_KYBER512, correct?

Reply:
You can use [DEFINE_POINTER_CLEANUP_FUNC](https://github.com/aws/s2n-tls/blob/main/utils/s2n_safety.h#L96) to define a new version of EVP_PKEY_CTX_free if it doesn't already exist. I'd also strongly suggest a unit test that checks that your feature is available in one environment you know it should be available in.

Reply:
ah, thank you for all the pointers. 

> I'd also strongly suggest a unit test that checks that your feature is available in one environment you know it should be available in.

so, if the `s2n_libcrypto_is_awslc()` runtime check is true, assert that they Kyber KEM interface is available?

Reply:
Whatever "environment" needs to be for this feature :) But I'd expect `s2n_libcrypto_is_awslc()` to be enough, if you expect Kyber to always be available if we're using awslc. Is it in all versions of awslc, or just recent ones?

You're not trying to capture all the cases where the interface would be available (we added feature detection to avoid that), just one as a sanity check.

Reply:
I added [this](https://github.com/aws/s2n-tls/blob/main/tests/features/evp_kyber_512.c) when we first migrated to Kyber in LC. Is this what you had in mind? If so, I think it might be prudent to extend it to check for some of the new KEM API interfaces.

Reply:
yep, i updated your feature test and renamed it:

features/evp_kyber_512.c → tests/features/evp_kem_kyber_512.c